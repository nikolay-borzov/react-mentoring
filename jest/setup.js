import promiseFinally from 'promise.prototype.finally'

import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

// Bring '.finally' to Promise (only for Node 8). Will be available in Node 10
promiseFinally.shim()

Enzyme.configure({ adapter: new Adapter() })

// Make Enzyme functions available in all test files without importing
global.shallow = shallow
global.render = render
global.mount = mount
