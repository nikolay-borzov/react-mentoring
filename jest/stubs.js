export const film = {
  id: 1,
  title: 'Film title',
  poster_path: 'http://via.placeholder.com/500x750',
  overview: 'Film overview',
  release_date: '2018-05-04',
  genres: ['some genre', 'another genre']
}

export const films = [film]
