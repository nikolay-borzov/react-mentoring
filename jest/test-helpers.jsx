import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectly = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectlyShallow = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

/**
 * Creates test closure to check whether component contains another component
 * @param {Function} getComponent Function returning parent component
 * @param {string} componentDisplayName Child component display name
 * @param {Object} params Optional params
 * @param {number} [params.testName=`contains ${componentDisplayName} component`]
 * @param {number} [params.expectedCount=1] Expected number of children components
 * @param {Object} params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent,
  componentDisplayName,
  { testName, expectedCount = 1, expectedProps } = {}
) => {
  it(testName || `contains ${componentDisplayName} component`, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = url => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga, state = {}, arg1 = undefined) => {
  const dispatched = []
  return runSagaOriginal(
    {
      dispatch: action => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
