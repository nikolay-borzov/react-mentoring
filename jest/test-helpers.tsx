import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'
import { AnyAction } from 'redux';

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 */
export const itRendersCorrectly = (
  getComponent: Function,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param getComponent Function returning component to test snapshot
 */
export const itRendersCorrectlyShallow = (
  getComponent: () => React.ReactElement,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

interface ItContainsComponentParams {
  testName?: string
  expectedCount?: number
  expectedProps?: {}
}

/**
 * Creates test closure to check whether component contains another component
 * @param getComponent Function returning parent component
 * @param componentDisplayName Child component display name
 * @param params Optional params
 * @param [params.testName=`contains ${componentDisplayName} component`]
 * @param [params.expectedCount=1] Expected number of children components
 * @param params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent: () => React.ReactElement,
  componentDisplayName: string,
  { testName = `contains ${componentDisplayName} component`, expectedCount = 1, expectedProps }: ItContainsComponentParams = {}
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = (url: string) => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga: any, state = {}, arg1: any = undefined) => {
  const dispatched: AnyAction[] = []
  return runSagaOriginal(
    {
      dispatch: (action: AnyAction) => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
