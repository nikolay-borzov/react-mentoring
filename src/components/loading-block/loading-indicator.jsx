// @flow

import React from 'react'
import styled, { keyframes } from 'styled-components'

const LoadingIndicatorStyled = styled.div`
  font-size: 250%;
`

/* Animation from https://github.com/tobiasahlin/SpinKit */
const LoadingSpinner = styled.div`
  text-align: center;
`

const bounce = keyframes`
  0%,
  80%,
  100% {
    transform: scale(0);
  }
  40% {
    transform: scale(1);
  }
`

const LoadingSpinnerBounce = styled.div`
  width: 1.5rem;
  height: 1.5rem;
  background-color: var(--color-default);

  border-radius: 100%;
  display: inline-block;
  animation: ${bounce} 1.4s infinite ease-in-out both;
  transform-origin: center center;

  .alt-background & {
    background-color: var(--color-text-light);
  }
`

const LoadingSpinnerBounceDelayedMore = LoadingSpinnerBounce.extend`
  animation-delay: -0.32s;
`

const LoadingSpinnerBounceDelayed = LoadingSpinnerBounce.extend`
  animation-delay: -0.16s;
`

export function LoadingIndicator({ hideText }: { hideText: boolean }) {
  return (
    <LoadingIndicatorStyled className="centered font-bold">
      <div>
        <LoadingSpinner>
          <LoadingSpinnerBounceDelayedMore />
          <LoadingSpinnerBounceDelayed />
          <LoadingSpinnerBounce />
        </LoadingSpinner>
        {hideText ? null : <React.Fragment>Loading&hellip;</React.Fragment>}
      </div>
    </LoadingIndicatorStyled>
  )
}
