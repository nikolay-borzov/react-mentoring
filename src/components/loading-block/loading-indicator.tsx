import React from 'react'
import styled, { keyframes } from 'styled-components'

const LoadingIndicatorStyled = styled.div`
  font-size: 250%;
`

/* Animation from https://github.com/tobiasahlin/SpinKit */
const LoadingSpinner = styled.div`
  text-align: center;
`

const bounce = keyframes`
  0%,
  80%,
  100% {
    transform: scale(0);
  }
  40% {
    transform: scale(1);
  }
`

const LoadingSpinnerBounce = styled.div`
  width: 1.5rem;
  height: 1.5rem;
  background-color: var(--color-default);

  border-radius: 100%;
  display: inline-block;
  animation: ${bounce} 1.4s infinite ease-in-out both;
  transform-origin: center center;

  .alt-background & {
    background-color: var(--color-text-light);
  }
`

const LoadingSpinnerBounceDelayedMore = styled(LoadingSpinnerBounce)`
  animation-delay: -0.32s;
`

const LoadingSpinnerBounceDelayed = styled(LoadingSpinnerBounce)`
  animation-delay: -0.16s;
`

interface LoadingIndicatorProps {
  hideText?: boolean
}

export const LoadingIndicator: React.FunctionComponent<
  LoadingIndicatorProps
> = ({ hideText }) => (
  <LoadingIndicatorStyled className="centered font-bold">
    <div>
      <LoadingSpinner>
        <LoadingSpinnerBounceDelayedMore />
        <LoadingSpinnerBounceDelayed />
        <LoadingSpinnerBounce />
      </LoadingSpinner>
      {hideText ? null : <>Loading&hellip;</>}
    </div>
  </LoadingIndicatorStyled>
)
