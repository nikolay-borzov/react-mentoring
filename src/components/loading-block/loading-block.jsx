// @flow

import * as React from 'react'

import { LoadingIndicator } from './loading-indicator'

type LoadingBlockProps = {
  isLoaded: boolean,
  children: React.ChildrenArray<React.Node>,
  hideText: boolean
}

LoadingBlock.defaultProps = {
  hideText: false
}

// TODO: Investigate why this is called four times (film-container.js) while props.isLoaded is changed only once
export function LoadingBlock(props: LoadingBlockProps) {
  if (props.isLoaded) {
    return props.children
  }

  return <LoadingIndicator hideText={props.hideText} />
}
