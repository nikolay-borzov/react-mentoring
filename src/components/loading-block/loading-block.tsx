import * as React from 'react'

import { LoadingIndicator } from './loading-indicator'

interface LoadingBlockProps {
  isLoaded: boolean
  children: React.ReactNode
  hideText?: boolean
}

// TODO: Investigate why this is called four times (film-container.js) while props.isLoaded is changed only once
export const LoadingBlock: React.FunctionComponent<
  LoadingBlockProps
> = props => {
  if (props.isLoaded) {
    return <>{props.children}</>
  }

  return <LoadingIndicator hideText={props.hideText} />
}

LoadingBlock.defaultProps = {
  hideText: false
}
