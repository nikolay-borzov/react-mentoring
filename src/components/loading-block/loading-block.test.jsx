import React from 'react'
import { shallow } from 'enzyme'

import { itRendersCorrectlyShallow } from '../../../jest/test-helpers'

import { LoadingBlock } from './loading-block'

describe('LoadingBlock component', () => {
  let props = {
    isLoaded: false
  }

  itRendersCorrectlyShallow(() => (
    <LoadingBlock {...props}>Content</LoadingBlock>
  ))

  it(`contains 'props.children' if 'isLoaded=true'`, () => {
    props.isLoaded = true

    const wrapper = shallow(
      <LoadingBlock {...props}>
        <div>Content</div>
      </LoadingBlock>
    )

    expect(wrapper.contains(<div>Content</div>)).toBe(true)
  })

  it(`renders 'LoadingIndicator' if 'isLoaded=false'`, () => {
    props.isLoaded = false

    const wrapper = shallow(
      <LoadingBlock {...props}>
        <div>Content</div>
      </LoadingBlock>
    )

    expect(wrapper.find('LoadingIndicator').length).toBe(1)
  })
})
