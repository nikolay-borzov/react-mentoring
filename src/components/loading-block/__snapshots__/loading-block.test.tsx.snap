// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingBlock component renders correctly 1`] = `
<LoadingIndicator
  hideText={false}
/>
`;
